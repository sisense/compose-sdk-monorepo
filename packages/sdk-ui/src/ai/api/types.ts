import { JaqlQueryPayload } from '@sisense/sdk-query-client';

import { ExpandedQueryModel } from '@/analytics-composer';
import { WidgetProps } from '@/props';

export interface ChatContext {
  title: string;
  live: boolean;
}

export interface ChatContextDetails {
  dashboardId?: string;
}
export interface TextMessage {
  role: 'assistant';
  content: string; // stringified json { answer: string }
  type: 'text';
}
export interface NlqMessage {
  role: 'assistant';
  content: string; // stringified json
  type: 'nlq';
}
interface RegularMessage {
  role: 'user' | 'assistant';
  content: string;
}
export type ChatMessage = RegularMessage | NlqMessage | TextMessage;

export interface Chat {
  chatId: string;
  contextId: string;
  contextTitle: string;
  contextDetails?: ChatContextDetails;
  expireAt: string;
  tenantId: string;
  userId: string;
  chatHistory: ChatMessage[];
}
export type ChatWithoutHistory = Omit<Chat, 'chatHistory'>;

export interface LlmConfig {
  provider: string;
  apiKey: string;
  baseUrl: string;
  model: string;
  version?: string;
}

export interface ChatRequest {
  text?: string;
  timezone?: string;
  options?: {
    enableFollowup?: boolean;
  };
}

interface TextResponse {
  chatHistory: ChatMessage[];
  data: {
    answer: string;
  };
  chatId: string;
  responseType: 'text';
}

interface ErrorResponse {
  data: {
    message: string;
    code: string;
    answer: string;
  };
  responseType: 'error';
}

/**
 * Response data for NLQ request
 */
export interface NlqResponseData extends ExpandedQueryModel {
  /** Detailed description of the response */
  detailedDescription: string;
  /** An array of suggested followup questions  */
  followupQuestions?: string[];
  /** NLQ prompt used in the request  */
  nlqPrompt: string;
  /** Suggested widget props*/
  widgetProps?: WidgetProps | undefined;
  /** The response message for the chat  */
  userMsg: string;
  /** @internal */
  clarification?: string;
}
export interface NlqResponse {
  data: NlqResponseData;
  timestamp: string;
  responseType: 'nlq';
}
export type ChatResponse = NlqResponse | TextResponse | ErrorResponse;

export interface QueryRecommendationConfig {
  numOfRecommendations: number;
  userPrompt?: string;
}

export type NlqResult = Omit<NlqResponseData, 'followupQuestions'>;

/** AI-generated recommended query you can run on your data model */
export interface QueryRecommendation extends Omit<NlqResponseData, 'followupQuestions'> {}
export type QueryRecommendationResponse = QueryRecommendation[];

export interface GetNlgInsightsRequest {
  jaql: Pick<JaqlQueryPayload, 'datasource' | 'metadata' | 'filterRelations'>;
  verbosity?: 'Low' | 'High';
}

/** @internal */
export interface GetNlgInsightsResponse {
  data?: {
    answer: string;
  };
  responseType: 'Text';
}

/** @internal */
export const NLQ_RESULT_CHART_TYPES = [
  'indicator',
  'column',
  'line',
  'table',
  'pie',
  'bar',
] as const;
/** NLQ result chart types @expandType */
export type NlqResultChartType = (typeof NLQ_RESULT_CHART_TYPES)[number];
export interface GetNlqResultRequest {
  text: string;
  timezone?: string;
  chartTypes: NlqResultChartType[];
}

export interface SendFeedbackRequest {
  type: string;
  data: object;
  sourceId: string;
  rating: -1 | 1;
}

/**
 * The chat mode to use for a chat session
 *
 * analyze: Enable business users to uncover data insights
 * develop: Enable developers to build queries and charts for embedded analytics
 */
export type ChatMode = 'analyze' | 'develop';
