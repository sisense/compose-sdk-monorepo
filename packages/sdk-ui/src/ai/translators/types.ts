import { MetadataItem } from '@sisense/sdk-query-client';
import { AxesMapping, NlqResponseData } from '@/ai';
import { ChartType, TableType } from '@/types';
import { WidgetModel } from '@/models';

/**
 * Expanded Query Model that is based on NlqResponseData.
 * It contains expanded JAQL and chart recommendations returned from the chat response.
 *
 * @internal
 */
export type ExpandedQueryModel = Pick<
  NlqResponseData,
  'jaql' | 'chartRecommendations' | 'queryTitle'
>;

/**
 * Expanded Query Model
 *
 * Used as a default value for ExpandedQueryModel
 *
 * @internal
 */
export const EMPTY_EXPANDED_QUERY_MODEL: ExpandedQueryModel = {
  jaql: { datasource: { title: '' }, metadata: [] },
  queryTitle: '',
  chartRecommendations: {},
};

/**
 * Simple Chart Recommendations
 *
 * @internal
 */
export interface SimpleChartRecommendations {
  chartType: string;
  dataOptions: AxesMapping;
}

/**
 * Simple Query Model that is a simplified version of ExpandedQueryModel.
 *
 * It removes any JAQL syntax or information that already exists in the schema.
 * Also, any parameters that have a default do not need to be specified.
 *
 * Users manipulates this model via the query in YAML format.
 *
 * ModelTranslator is responsible for translating this query model to SimpleWidgetModel.
 *
 * @internal
 */
export interface SimpleQueryModel {
  /**
   * The data model or perspective title.
   */
  model: string;
  metadata: MetadataItem[];
  chart: SimpleChartRecommendations | {};
  /**
   * Title of the query
   */
  queryTitle: string;
}

/**
 * Empty Simple Query Model
 *
 * Used as a default value for SimpleQueryModel
 *
 * @internal
 */
export const EMPTY_SIMPLE_QUERY_MODEL: SimpleQueryModel = {
  model: '',
  metadata: [],
  chart: {},
  queryTitle: '',
};

/**
 * All Chart Types including Table
 * Note: Table will be re-united with ChartType in a future MR
 *
 * @internal
 */
export type AllChartType = ChartType | TableType;

/**
 * Chart Widget Model
 *
 * This model serves two purposes:
 *
 * (1) For rendering a chart or table widget in the Query Composer.
 * (2) For reverse engineered (translated) to CSDK code
 *
 * Naming TBD
 *
 * @internal
 */
export type ChartWidgetModel = Pick<WidgetModel, 'getChartWidgetProps' | 'getTableWidgetProps'> & {
  chartType: AllChartType;
  queryTitle: string;
};

/**
 * UI Framework
 *
 * @internal
 */
export type UiFramework = 'react' | 'vue' | 'angular';

/**
 * Code Template Key
 *
 * @internal
 */
export type CodeTemplateKey = 'baseChartTmpl' | 'chartTmpl' | 'chartWidgetTmpl' | 'tableWidgetTmpl';

/**
 * Code Templates
 *
 * @internal
 */
export type CodeTemplates = {
  [key in UiFramework]: { [key in CodeTemplateKey]: string };
};
