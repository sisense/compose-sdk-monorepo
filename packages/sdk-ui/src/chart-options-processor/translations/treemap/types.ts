import { Chart, PointLabelObject } from '@sisense/sisense-charts';

export type ParentValues = {
  direction: number;
  height: number;
  val: number;
  width: number;
  x: number;
  y: number;
};

export type LayoutPointResult = {
  x: number;
  y: number;
  width: number;
  height: number;
};

export type TreemapLayoutAlgorithmContext = {
  chart: Chart;
  squarified: (parent: ParentValues, children: PointLabelObject[]) => LayoutPointResult[];
};
