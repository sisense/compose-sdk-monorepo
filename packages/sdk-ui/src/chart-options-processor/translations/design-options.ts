import { CalendarDayOfWeek } from '@/chart/restructured-charts/highchart-based-charts/calendar-heatmap-chart/utils';

import {
  CalendarHeatmapViewType,
  Convolution,
  LegendOptions,
  LineOptions,
  ScattermapMarkers,
  SeriesLabels,
  SunburstStyleOptions,
  TotalLabels,
  TreemapStyleOptions,
} from '../../types';
import { Axis } from './axis-section';
import { FunnelDirection, FunnelLabels, FunnelSize, FunnelType } from './funnel-plot-options';
import { Marker } from './marker-section';
import { PieLabels, PieType } from './pie-plot-options';
import { ScatterMarkerSize } from './scatter-plot-options';
import { LineType, StackType } from './translations-to-highcharts';
import { DesignPerSeries, TextStyle } from './types';

type DataLimits = {
  seriesCapacity: number;
  categoriesCapacity: number;
};

export type BaseDesignOptionsType = {
  legend?: LegendOptions;
  seriesLabels?: SeriesLabels;
  lineType: LineType;
  /**
   * @deprecated
   * Use line.width instead
   */
  lineWidth: number;
  marker: Marker;
  xAxis: Axis;
  yAxis: Axis;
  x2Axis?: Axis;
  y2Axis?: Axis;
  autoZoom: {
    enabled: boolean;
    scrollerLocation?: {
      min: number;
      max: number;
    };
  };
  dataLimits: DataLimits;
};

export type CartesianChartDesignOptions = BaseDesignOptionsType & DesignPerSeries;
export type StackableChartDesignOptions = CartesianChartDesignOptions & {
  line?: LineOptions;
  stackType: StackType;
  totalLabels?: TotalLabels;
  itemPadding?: number;
  groupPadding?: number;
  borderRadius?: number | string;
};

export type LineChartDesignOptions = CartesianChartDesignOptions & {
  /** Step type for step line charts: left, center, or right */
  step?: 'left' | 'center' | 'right';
  line?: LineOptions;
};

export type AreaChartDesignOptions = StackableChartDesignOptions;

export type BarChartDesignOptions = StackableChartDesignOptions;

export type ColumnChartDesignOptions = StackableChartDesignOptions;

// todo: remove BaseDesignOptionsType after refactor of `ChartDesignOptions` usage
export type CalendarHeatmapChartDesignOptions = BaseDesignOptionsType & {
  width?: number;
  height?: number;
  viewType: CalendarHeatmapViewType;
  cellSize?: number;
  startOfWeek: CalendarDayOfWeek;
  cellLabels: {
    enabled: boolean;
    style?: TextStyle;
  };
  dayLabels: {
    enabled: boolean;
    style?: TextStyle;
  };
  monthLabels: {
    enabled: boolean;
    style?: TextStyle;
  };
  weekends: {
    enabled: boolean;
    days: CalendarDayOfWeek[];
    cellColor?: string;
    hideValues: boolean;
  };
  pagination: {
    enabled: boolean;
    style?: TextStyle;
    startMonth?: Date;
  };
};

export type PolarType = 'line' | 'area' | 'column';
export type PolarChartDesignOptions = CartesianChartDesignOptions & {
  polarType: PolarType;
  totalLabels?: TotalLabels;
};

export function isPolarChartDesignOptions(
  options: CartesianChartDesignOptions,
): options is PolarChartDesignOptions {
  return 'polarType' in options;
}

export type PieChartDesignOptions = BaseDesignOptionsType & {
  pieType?: PieType;
  pieLabels?: PieLabels;
  convolution?: Convolution;
};

export type FunnelChartDesignOptions = BaseDesignOptionsType & {
  funnelSize?: FunnelSize;
  funnelType?: FunnelType;
  funnelDirection?: FunnelDirection;
  funnelLabels?: FunnelLabels;
};

export type TreemapChartDesignOptions = BaseDesignOptionsType & TreemapStyleOptions;

export type SunburstChartDesignOptions = BaseDesignOptionsType & SunburstStyleOptions;

export type ScatterChartDesignOptions = BaseDesignOptionsType & {
  markerSize?: ScatterMarkerSize;
};

/**
 * Configuration for Table color options
 */
export type TableColorOptions = {
  /**
   * Boolean flag whether to fill with color
   */
  enabled: boolean;
  /**
   * Color used for background fill
   */
  backgroundColor?: string;
  /**
   * Color used for text
   */
  textColor?: string;
};

/**
 * Configuration for Table design options
 *
 */
export type TableDesignOptions = {
  /**
   * Vertical padding around whole table
   * Default value is 20
   *
   */
  paddingVertical?: number;
  /**
   * Horizontal padding around whole table
   * Default value is 20
   *
   */
  paddingHorizontal?: number;
  /**
   * Header options
   */
  header?: {
    /**
     * Color of header
     */
    color?: TableColorOptions;
  };
  /**
   * Columns options
   */
  columns?: {
    /**
     * Alternating color for columns
     */
    alternatingColor?: TableColorOptions;
    /**
     * Modes of columns width
     * 'auto' - all columns will have the same width and fit the table width (no horizontal scroll)
     * 'content' - columns width will be based on content (default option)
     */
    width?: 'auto' | 'content';
  };
  /**
   * Rows options
   */
  rows?: {
    /**
     * Alternating color for rows
     */
    alternatingColor?: TableColorOptions;
  };
};

export type IndicatorStyleType = 'numeric' | 'gauge';
export type NumericIndicatorSubType = 'numericSimple' | 'numericBar';
export type IndicatorSkin = 'vertical' | 'horizontal' | 1 | 2;

/** Configuration options that define components of an indicator chart. */
export type IndicatorComponents = {
  /** The main title of the indicator chart */
  title?: {
    /** Whether the title should be shown */
    shouldBeShown: boolean;
    /** The text of the title */
    text?: string;
  };
  /** The secondary title of the indicator chart to be shown when `secondary` is specified in {@link IndicatorChartDataOptions} */
  secondaryTitle?: {
    /** The text of the secondary title */
    text?: string;
  };
  /** The ticks displayed on the indicator chart */
  ticks?: {
    /** Whether the ticks should be shown */
    shouldBeShown: boolean;
  };
  /** The value labels of the indicator chart */
  labels?: {
    /** Whether the labels should be shown */
    shouldBeShown: boolean;
  };
};

/**
 * Design options for an indicator chart.
 */
export type IndicatorChartDesignOptions<
  IndicatorType extends IndicatorStyleType = IndicatorStyleType,
> = BaseDesignOptionsType & {
  indicatorComponents: IndicatorComponents;
  forceTickerView: boolean;
} & (IndicatorType extends 'gauge' ? GaugeSpecificDesignOptions : NumericSpecificDesignOptions);

export type GaugeSpecificDesignOptions = {
  indicatorType: 'gauge';
  skin: 1 | 2;
  tickerBarHeight?: number;
};

export type NumericSpecificDesignOptions<
  NumericSubtype extends NumericIndicatorSubType = NumericIndicatorSubType,
> = {
  indicatorType: 'numeric';
  numericSubtype: NumericSubtype;
  skin: NumericSubtype extends 'numericSimple' ? 'vertical' | 'horizontal' : never;
};

export type BoxplotType = 'full' | 'hollow';
export type BoxplotChartDesignOptions = BaseDesignOptionsType & {
  boxplotType: BoxplotType;
};

// todo: remove `BaseDesignOptionsType` after refactor of `ChartDesignOptions` usage
export type ScattermapChartDesignOptions = BaseDesignOptionsType & {
  markers: Required<ScattermapMarkers>;
};

export type AreaRangeChartDesignOptions = BaseDesignOptionsType & {
  /** Configuration that defines line style */
  line?: LineOptions;
};
