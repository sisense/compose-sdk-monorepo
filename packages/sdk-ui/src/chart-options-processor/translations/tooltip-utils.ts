import { isNumber } from '@sisense/sdk-data';
import { Category, Value, isValue } from '../../chart-data-options/types.js';
import { applyFormat, getCompleteNumberFormatConfig } from './number-format-config.js';
import type { SeriesChartType } from '@/types';

export const isXValueNumeric = (dataOptionX: Value | Category | undefined) =>
  dataOptionX ? isValue(dataOptionX) || (dataOptionX.type && isNumber(dataOptionX.type)) : false;

export const formatTooltipValue = (
  dataOption: Value | Category | undefined,
  value: number | undefined,
  displayValue: string,
) => {
  if (!dataOption || value === undefined || isNaN(value)) return displayValue;

  const numberFormatConfig = getCompleteNumberFormatConfig(dataOption.numberFormatConfig);
  return applyFormat(numberFormatConfig, value);
};

export const formatTooltipXValue = (
  dataOption: Value | Category | undefined,
  value: number | string | undefined,
  displayValue: string,
) => {
  const isNumeric = isXValueNumeric(dataOption);

  return isNumeric
    ? formatTooltipValue(dataOption, parseFloat(`${value}`), displayValue)
    : displayValue;
};

const percentSupportedSubChartTypes: SeriesChartType[] = [
  'auto',
  'column',
  'area',
  'areaspline',
  'bar',
  'scatter',
];
export function isTooltipPercentValueSupported(options: Value | undefined) {
  if (options?.chartType) {
    return percentSupportedSubChartTypes.includes(options.chartType);
  }
  return true;
}

export type TooltipSettings = {
  enabled?: boolean;
  animation?: boolean;
  backgroundColor?: string;
  borderColor?: string;
  borderRadius?: number;
  borderWidth?: number;
  useHTML?: boolean;
  formatter?: () => string;
  style?: {
    fontFamily?: string;
  };
};

export type InternalSeries = {
  series: { name: string; color: string };
  x: string;
  y: number;
  point: {
    x?: number | string;
    y?: number;
    z?: number;
    high?: number;
    low?: number;
    q1?: number;
    q3?: number;
    median?: number;
    name: string;
    color: string;
    custom?: {
      number1?: number;
      string1?: string;
      xDisplayValue?: string;
      xValue?: (number | string)[];
    };
  };
  percentage?: number;
};
