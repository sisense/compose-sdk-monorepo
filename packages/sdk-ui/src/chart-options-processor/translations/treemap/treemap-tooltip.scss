.csdk-treemap-tooltip-wrapper {
  min-width: 210px;
  color: #acacac;
  font-size: 16px;
  line-height: 16px;
  background: white;
  border-radius: 10px;
  overflow: hidden;
  margin-right: -1px;
  padding-bottom: 10px;
}

.csdk-treemap-tooltip-level {
  display: flex;
  justify-content: space-between;
  position: relative;
  padding: 8px;
  font-size: 16px;
  border-bottom: 1px solid #ccc;

  span {
    padding: 0 2px;
  }
}

.csdk-treemap-tooltip-value {
  padding: 10px 10px 0;
  font-size: 12px;
  line-height: 16px;
}
