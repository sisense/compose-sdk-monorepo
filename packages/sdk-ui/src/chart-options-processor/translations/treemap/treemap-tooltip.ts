/* eslint-disable max-lines-per-function */
/* eslint-disable sonarjs/cognitive-complexity */
/* eslint-disable max-lines */
import { CategoricalChartDataOptionsInternal } from '../../../chart-data-options/types';
import { TreemapChartDesignOptions } from '../design-options';
import { InternalSeriesNode, TooltipSettings } from '../../tooltip';
import { colorChineseSilver, colorWhite } from '../../../chart-data-options/coloring/consts';
import { applyFormat, defaultConfig } from '../number-format-config';
import { TooltipFormatterContextObject } from '@sisense/sisense-charts';
import './treemap-tooltip.scss';

type TooltipFormatterOptions = {
  displayTotalContribution: boolean;
  displayColorCircles: boolean;
  shouldSkip: (context: TooltipFormatterContextObject) => boolean;
};

export const getTreemapTooltipSettings = (
  chartDataOptions: CategoricalChartDataOptionsInternal,
  designOptions: TreemapChartDesignOptions,
  formatterOptions?: TooltipFormatterOptions,
): TooltipSettings => ({
  animation: false,
  backgroundColor: colorWhite,
  borderColor: colorChineseSilver,
  borderRadius: 10,
  borderWidth: 1,
  padding: 1,
  useHTML: true,
  outside: true,
  formatter: function (this) {
    return treemapTooltipFormatter(this, chartDataOptions, designOptions, formatterOptions);
  },
});

const triangle = `
  <div style="position: absolute; top: 100%; right: 40px; z-index: 1">
    <div style="
      border: solid transparent;
      content: ' ';
      height: 0;
      width: 0;
      position: absolute;
      pointer-events: none;
      border-color: rgba(136,136,136,0);
      border-top-color: #ccc;
      border-width: 7px 12px 12px 12px;
      margin-left: -7px;
    "></div>
     <div style="
       border: solid transparent;
       content: ' ';
       height: 0;
       width: 0;
       position: absolute;
       pointer-events: none;
       border-color: rgba(255,255,255,0);
       border-top-color: #fff;
       border-width: 6px 11px 11px 11px;
       margin-left: -6px;
     "></div>
  </div>
`;

function getContribution(value: number, total: number) {
  return value / (total / 100);
}

function getFormattedContribution(value: number, total: number) {
  return `${getContribution(value, total).toPrecision(3)}%`;
}

export function treemapTooltipFormatter(
  context: TooltipFormatterContextObject,
  chartDataOptions: CategoricalChartDataOptionsInternal,
  designOptions: TreemapChartDesignOptions,
  formatterOptions?: TooltipFormatterOptions,
) {
  if (formatterOptions?.shouldSkip && formatterOptions.shouldSkip(context)) {
    return false;
  }
  const numberFormatConfig = chartDataOptions.y?.[0]?.numberFormatConfig ?? defaultConfig;
  const isContributionMode = designOptions?.tooltip?.mode === 'contribution';
  const valueTitle = chartDataOptions.y?.[0]?.title ?? chartDataOptions.y?.[0]?.name;
  const color = context.color as string;

  let rootValue = 0;
  const nodesToShow: InternalSeriesNode[] = [];
  const handleNode = (node: InternalSeriesNode) => {
    if (node.parentNode) {
      nodesToShow.push(node);
    } else {
      rootValue = node.val;
      return;
    }
    handleNode(node.parentNode);
  };
  // eslint-disable-next-line
  handleNode(context.point['node'] as InternalSeriesNode);

  return `
      <div
        class="csdk-treemap-tooltip-wrapper">
        ${[...nodesToShow]
          .reverse()
          .map((node, index) => {
            const isLastNode = index === nodesToShow.length - 1;
            return `
                <div
                  class="csdk-treemap-tooltip-level"
                  style="background: ${isLastNode ? '#f2f2f2' : 'white'};"
                  >
                  ${
                    formatterOptions?.displayColorCircles
                      ? prepareTooltipColorCircle(node.color as string)
                      : ''
                  }
                  <span
                    style="color: ${isLastNode ? color : 'currentColor'}"
                  >${node.name}</span>
                  <span
                    style="color: ${isLastNode ? '#5B6372' : 'currentColor'}; margin-left: auto;"
                  >${
                    isContributionMode
                      ? getFormattedContribution(node.val, rootValue)
                      : applyFormat(numberFormatConfig, node.val)
                  }</span>
                  ${isLastNode ? '' : triangle}
                </div>`;
          })
          .join('')}
        <div class="csdk-treemap-tooltip-value">
          <span style="color:${color}; font-size: 15px">${
    isContributionMode
      ? getFormattedContribution(nodesToShow[0].val, rootValue)
      : applyFormat(numberFormatConfig, nodesToShow[0]?.val)
  }</span> of ${valueTitle}
        </div>
        ${
          nodesToShow[1]
            ? `<div class="csdk-treemap-tooltip-value" style="padding-top: 4px;">
                  <span style="color:${color}; font-size: 15px">${getFormattedContribution(
                nodesToShow[0]?.val,
                nodesToShow[1]?.val,
              )}
                  </span> of ${nodesToShow[1]?.name}
                </div>`
            : ''
        }
        ${
          formatterOptions?.displayTotalContribution && !isContributionMode
            ? `<div class="csdk-treemap-tooltip-value" style="padding-top: 4px;">
                    <span style="color:${color}; font-size: 15px">${getFormattedContribution(
                nodesToShow[0]?.val,
                rootValue,
              )}
                    </span> of total
                   </div>`
            : ''
        }
        </div>
      </div>`;
}

function prepareTooltipColorCircle(color: string) {
  return `
    <div style="
      width: 12px;
      height: 12px;
      border-radius: 100%;
      background-color: ${color};
      margin-right: 6px;
    "></div>
  `;
}
