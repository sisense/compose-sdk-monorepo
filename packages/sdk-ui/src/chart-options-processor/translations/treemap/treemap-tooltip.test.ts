import { TFunction } from '@sisense/sdk-common';
import get from 'lodash/get';

import { CategoricalChartDataOptionsInternal } from '../../../chart-data-options/types';
import { translation as enTranslation } from '../../../translation/resources/en';
import { TreemapChartDesignOptions } from '../design-options';
import { HighchartsDataPointContext } from '../tooltip-utils';
import { treemapTooltipFormatter } from './treemap-tooltip';

const translateMock = ((path: string) => {
  return get(enTranslation, path, '');
}) as TFunction;

describe('Treemap Chart tooltip formatter', () => {
  const dataOptions = {
    y: [{ column: { title: 'test' } }],
  } as CategoricalChartDataOptionsInternal;

  const designOptions = {} as TreemapChartDesignOptions;

  const node1 = {
    val: 1,
    name: 'One',
  };
  const node2 = {
    val: 2,
    name: 'Two',
  };
  const node3 = {
    val: 3,
    name: 'Three',
  };
  const node4 = {
    val: 4,
    name: 'Four',
  };

  it('single category tooltip', () => {
    const context = {
      color: 'red',
      point: {
        node: {
          ...node1,
          parentNode: node2,
        },
      },
    } as unknown as HighchartsDataPointContext;

    expect(
      treemapTooltipFormatter(context, dataOptions, designOptions, translateMock),
    ).toMatchSnapshot();
  });

  it('three categories tooltip', () => {
    const context = {
      color: 'blue',
      point: {
        node: {
          ...node1,
          parentNode: {
            ...node2,
            parentNode: {
              ...node3,
              parentNode: {
                ...node4,
              },
            },
          },
        },
      },
    } as unknown as HighchartsDataPointContext;

    expect(
      treemapTooltipFormatter(context, dataOptions, designOptions, translateMock),
    ).toMatchSnapshot();
  });

  it('contribution mode', () => {
    const context = {
      color: 'blue',
      point: {
        node: {
          ...node1,
          parentNode: {
            ...node2,
            parentNode: {
              ...node3,
              parentNode: {
                ...node4,
              },
            },
          },
        },
      },
    } as unknown as HighchartsDataPointContext;
    const designOptionsWithContribution = {
      tooltip: { mode: 'contribution' },
    } as TreemapChartDesignOptions;

    expect(
      treemapTooltipFormatter(context, dataOptions, designOptionsWithContribution, translateMock),
    ).toMatchSnapshot();
  });
});
