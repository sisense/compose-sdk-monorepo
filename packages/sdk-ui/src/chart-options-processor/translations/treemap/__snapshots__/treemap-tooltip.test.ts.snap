// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Treemap Chart tooltip formatter > contribution mode 1`] = `
"
      <div
        class="csdk-treemap-tooltip-wrapper">

                <div
                  class="csdk-treemap-tooltip-level"
                  style="background: white;"
                  >

                  <span
                    style="color: currentColor"
                  >Three</span>
                  <span
                    style="color: currentColor; margin-left: auto;"
                  >75.0%</span>

  <div style="position: absolute; top: 100%; right: 40px; z-index: 1">
    <div style="
      border: solid transparent;
      content: ' ';
      height: 0;
      width: 0;
      position: absolute;
      pointer-events: none;
      border-color: rgba(136,136,136,0);
      border-top-color: #ccc;
      border-width: 7px 12px 12px 12px;
      margin-left: -7px;
    "></div>
     <div style="
       border: solid transparent;
       content: ' ';
       height: 0;
       width: 0;
       position: absolute;
       pointer-events: none;
       border-color: rgba(255,255,255,0);
       border-top-color: #fff;
       border-width: 6px 11px 11px 11px;
       margin-left: -6px;
     "></div>
  </div>

                </div>
                <div
                  class="csdk-treemap-tooltip-level"
                  style="background: white;"
                  >

                  <span
                    style="color: currentColor"
                  >Two</span>
                  <span
                    style="color: currentColor; margin-left: auto;"
                  >50.0%</span>

  <div style="position: absolute; top: 100%; right: 40px; z-index: 1">
    <div style="
      border: solid transparent;
      content: ' ';
      height: 0;
      width: 0;
      position: absolute;
      pointer-events: none;
      border-color: rgba(136,136,136,0);
      border-top-color: #ccc;
      border-width: 7px 12px 12px 12px;
      margin-left: -7px;
    "></div>
     <div style="
       border: solid transparent;
       content: ' ';
       height: 0;
       width: 0;
       position: absolute;
       pointer-events: none;
       border-color: rgba(255,255,255,0);
       border-top-color: #fff;
       border-width: 6px 11px 11px 11px;
       margin-left: -6px;
     "></div>
  </div>

                </div>
                <div
                  class="csdk-treemap-tooltip-level"
                  style="background: #f2f2f2;"
                  >

                  <span
                    style="color: blue"
                  >One</span>
                  <span
                    style="color: #5B6372; margin-left: auto;"
                  >25.0%</span>

                </div>
        <div class="csdk-treemap-tooltip-value">
          <span style="color:blue; font-size: 15px">25.0%</span> of test
        </div>
        <div class="csdk-treemap-tooltip-value" style="padding-top: 4px;">
                  <span style="color:blue; font-size: 15px">50.0%
                  </span> of Two
                </div>

        </div>
      </div>"
`;

exports[`Treemap Chart tooltip formatter > single category tooltip 1`] = `
"
      <div
        class="csdk-treemap-tooltip-wrapper">

                <div
                  class="csdk-treemap-tooltip-level"
                  style="background: #f2f2f2;"
                  >

                  <span
                    style="color: red"
                  >One</span>
                  <span
                    style="color: #5B6372; margin-left: auto;"
                  >1</span>

                </div>
        <div class="csdk-treemap-tooltip-value">
          <span style="color:red; font-size: 15px">1</span> of test
        </div>


        </div>
      </div>"
`;

exports[`Treemap Chart tooltip formatter > three categories tooltip 1`] = `
"
      <div
        class="csdk-treemap-tooltip-wrapper">

                <div
                  class="csdk-treemap-tooltip-level"
                  style="background: white;"
                  >

                  <span
                    style="color: currentColor"
                  >Three</span>
                  <span
                    style="color: currentColor; margin-left: auto;"
                  >3</span>

  <div style="position: absolute; top: 100%; right: 40px; z-index: 1">
    <div style="
      border: solid transparent;
      content: ' ';
      height: 0;
      width: 0;
      position: absolute;
      pointer-events: none;
      border-color: rgba(136,136,136,0);
      border-top-color: #ccc;
      border-width: 7px 12px 12px 12px;
      margin-left: -7px;
    "></div>
     <div style="
       border: solid transparent;
       content: ' ';
       height: 0;
       width: 0;
       position: absolute;
       pointer-events: none;
       border-color: rgba(255,255,255,0);
       border-top-color: #fff;
       border-width: 6px 11px 11px 11px;
       margin-left: -6px;
     "></div>
  </div>

                </div>
                <div
                  class="csdk-treemap-tooltip-level"
                  style="background: white;"
                  >

                  <span
                    style="color: currentColor"
                  >Two</span>
                  <span
                    style="color: currentColor; margin-left: auto;"
                  >2</span>

  <div style="position: absolute; top: 100%; right: 40px; z-index: 1">
    <div style="
      border: solid transparent;
      content: ' ';
      height: 0;
      width: 0;
      position: absolute;
      pointer-events: none;
      border-color: rgba(136,136,136,0);
      border-top-color: #ccc;
      border-width: 7px 12px 12px 12px;
      margin-left: -7px;
    "></div>
     <div style="
       border: solid transparent;
       content: ' ';
       height: 0;
       width: 0;
       position: absolute;
       pointer-events: none;
       border-color: rgba(255,255,255,0);
       border-top-color: #fff;
       border-width: 6px 11px 11px 11px;
       margin-left: -6px;
     "></div>
  </div>

                </div>
                <div
                  class="csdk-treemap-tooltip-level"
                  style="background: #f2f2f2;"
                  >

                  <span
                    style="color: blue"
                  >One</span>
                  <span
                    style="color: #5B6372; margin-left: auto;"
                  >1</span>

                </div>
        <div class="csdk-treemap-tooltip-value">
          <span style="color:blue; font-size: 15px">1</span> of test
        </div>
        <div class="csdk-treemap-tooltip-value" style="padding-top: 4px;">
                  <span style="color:blue; font-size: 15px">50.0%
                  </span> of Two
                </div>

        </div>
      </div>"
`;
